src/binaryen-c.h

```diff
@@ -0,0 +1,246 @@
+
+struct BinaryenLiteral BinaryenLiteralInt32(int32_t x);
+struct BinaryenLiteral BinaryenLiteralInt64(int64_t x);
+struct BinaryenLiteral BinaryenLiteralFloat32(float x);
+struct BinaryenLiteral BinaryenLiteralFloat64(double x);
+struct BinaryenLiteral BinaryenLiteralFloat32Bits(int32_t x);
+struct BinaryenLiteral BinaryenLiteralFloat64Bits(int64_t x);
+
+// Expressions
+//
+// Some expressions have a BinaryenOp, which is the more
+// specific operation/opcode.
+//
+// Some expressions have optional parameters, like Return may not
+// return a value. You can supply a NULL pointer in those cases.
+
+typedef int32_t BinaryenOp;
+
+BinaryenOp BinaryenClz(void);
+BinaryenOp BinaryenCtz(void);
+BinaryenOp BinaryenPopcnt(void);
+BinaryenOp BinaryenNeg(void);
+BinaryenOp BinaryenAbs(void);
+BinaryenOp BinaryenCeil(void);
+BinaryenOp BinaryenFloor(void);
+BinaryenOp BinaryenTrunc(void);
+BinaryenOp BinaryenNearest(void);
+BinaryenOp BinaryenSqrt(void);
+BinaryenOp BinaryenEqz(void);
+BinaryenOp BinaryenExtendSInt32(void);
+BinaryenOp BinaryenExtentUInt32(void);
+BinaryenOp BinaryenWrapInt64(void);
+BinaryenOp BinaryenTruncSFloat32(void);
+BinaryenOp BinaryenTruncUFloat32(void);
+BinaryenOp BinaryenTruncSFloat64(void);
+BinaryenOp BinaryenTruncUFloat64(void);
+BinaryenOp BinaryenReinterpretFloat(void);
+BinaryenOp BinaryenConvertSInt32(void);
+BinaryenOp BinaryenConvertUInt32(void);
+BinaryenOp BinaryenConvertSInt64(void);
+BinaryenOp BinaryenConvertUInt64(void);
+BinaryenOp BinaryenPromoteFloat32(void);
+BinaryenOp BinaryenDemoteFloat64(void);
+BinaryenOp BinaryenReinterpretInt(void);
+BinaryenOp BinaryenAdd(void);
+BinaryenOp BinaryenSub(void);
+BinaryenOp BinaryenMul(void);
+BinaryenOp BinaryenDivS(void);
+BinaryenOp BinaryenDivU(void);
+BinaryenOp BinaryenRemS(void);
+BinaryenOp BinaryenRemU(void);
+BinaryenOp BinaryenAnd(void);
+BinaryenOp BinaryenOr(void);
+BinaryenOp BinaryenXor(void);
+BinaryenOp BinaryenShl(void);
+BinaryenOp BinaryenShrU(void);
+BinaryenOp BinaryenShrS(void);
+BinaryenOp BinaryenRotL(void);
+BinaryenOp BinaryenRotR(void);
+BinaryenOp BinaryenDiv(void);
+BinaryenOp BinaryenCopySign(void);
+BinaryenOp BinaryenMin(void);
+BinaryenOp BinaryenMax(void);
+BinaryenOp BinaryenEq(void);
+BinaryenOp BinaryenNe(void);
+BinaryenOp BinaryenLtS(void);
+BinaryenOp BinaryenLtU(void);
+BinaryenOp BinaryenLeS(void);
+BinaryenOp BinaryenLeU(void);
+BinaryenOp BinaryenGtS(void);
+BinaryenOp BinaryenGtU(void);
+BinaryenOp BinaryenGes(void);
+BinaryenOp BinaryenGeU(void);
+BinaryenOp BinaryenLt(void);
+BinaryenOp BinaryenLe(void);
+BinaryenOp BinaryenGt(void);
+BinaryenOp BinaryenGe(void);
+BinaryenOp BinaryenPageSize(void);
+BinaryenOp BinaryenCurrentMemory(void);
+BinaryenOp BinaryenGrowMemory(void);
+BinaryenOp BinaryenHasFeature(void);
+
+typedef void* BinaryenExpressionRef;
+
+// Block: name can be NULL
+BinaryenExpressionRef BinaryenBlock(BinaryenModuleRef module, const char* name, BinaryenExpressionRef* children, BinaryenIndex numChildren);
```

Comment:
Interesting point. I think you're right and we should probably have both. Although then it seems we'd also want an API to set the size of the basic block before incrementally adding to it (otherwise the vector contents would get reallocated)?